### PR TITLE
Add `force` option to annotation sync job

### PR DESCRIPTION
**Pull request by seanh**
If `force` is `True` then it both (re-)indexes the annotation (regardless of whether it's already present or up to date in Elasticsearch) and also immediately deletes the job from the queue.

### CODE REVIEW

File: h/services/search_index/_queue.py [hunk 1]
```diff
@@ -93,7 +99,9 @@ def sync(self, limit):
             return
 
         annotation_ids = {
-            URLSafeUUID.hex_to_url_safe(job.kwargs["annotation_id"]) for job in jobs
+            URLSafeUUID.hex_to_url_safe(job.kwargs["annotation_id"])
+            for job in jobs
+            if not job.kwargs.get("force", False)
```
Comment (seanh) [inline h/services/search_index/_queue.py:104]: There's no point in reading the annotation from Postgres or Elastic if `force` is `True`, we aren't going to use it

File: tests/h/services/search_index/_queue_test.py [hunk 2]
```diff
@@ -36,30 +38,48 @@ def test_it(self, db_session, factories, queue, now):
...
+                                        URLSafeUUID.url_safe_to_hex(annotation_id)
+                                    )
+                                ),
+                                "force": force,
+                            },
+                        }
+                    )
+                    for annotation_id in annotation_ids
+                ]
+            ).only()
```
Comment (seanh) [inline tests/h/services/search_index/_queue_test.py:81]: Giving h-matchers a workout here. `URLSafeUUID.url_safe_to_hex()` actually doesn't return strings that match those in the DB. There's further processing that happens after that because the DB is using a UUID type. So you have to do `str(uuid.UUID(URLSafeUUID.url_safe_to_hex(annotation_id)))`

File: h/services/search_index/_queue.py [hunk 3]
```diff
@@ -93,10 +99,16 @@ def sync(self, limit):
...
         }
-        annotations_from_db = self._get_annotations_from_db(annotation_ids)
-        annotations_from_es = self._get_annotations_from_es(annotation_ids)
+        if annotation_ids:
+            annotations_from_db = self._get_annotations_from_db(annotation_ids)
+            annotations_from_es = self._get_annotations_from_es(annotation_ids)
+        else:
+            annotations_from_db = {}
+            annotations_from_es = {}
```
Comment (seanh) [inline h/services/search_index/_queue.py:111]: There's no point in reading the annotation from Postgres or Elastic if `force` is `True`, we aren't going to use it. If all of the jobs have `force=True` then there's no point in making a Postgres or Elastic query at all.

Comment (seanh) [inline h/services/search_index/_queue.py:111]: There's no point in reading the annotation from Postgres or Elastic if `force` is `True`, we aren't going to use it. If all of the jobs have `force=True` then there's no point in making a Postgres or Elastic query at all.

File: tests/h/services/search_index/_queue_test.py (hunk 2 above)
Comment (jon-betts) [reply tests/h/services/search_index/_queue_test.py:81]: Oh good gravy.

These comparisons get pretty long and rambling, but they still read ok mostly.

File: h/services/search_index/_queue.py [hunk 4]
```diff
@@ -113,7 +125,11 @@ def sync(self, limit):
             annotation_from_db = annotations_from_db.get(annotation_id)
             annotation_from_es = annotations_from_es.get(annotation_id)
 
-            if not annotation_from_db:
+            if job.kwargs.get("force", False):
```
Comment (jon-betts) [inline h/services/search_index/_queue.py:128]: :+1: Use of `get()` is very JSON-positive

File: h/services/search_index/_queue.py (hunk 3 above)
Comment (jon-betts) [reply h/services/search_index/_queue.py:111]: This actually fits together really nicely. I'm not sure why I thought it would be a good idea to treat them differently.

The only reason I can think is that the performance might be quite different. On the whole I bet we don't issue many writes when doing the normal run of the mill, but we could go absolutely crackers in the case of a big forced index.

This is actually quite easy to remedy if we ever found it to be a real problem. We could have a separate limit for writes. 

So for example, we could do our scan of 1000, find that all 1000 want to be written, know from experience it's only going to work in time with 750 say, and then just bail out. The next iteration would pick the jobs up (because we wouldn't mark them as done).

Long story short, I doubt it matters unless we end up cranking the number of reads up a lot (say we started doing 10,000 at a time because it was very efficient, and we expect to do an average of 1 write) and it's easy to fix if we want to.

File: tests/h/services/search_index/_queue_test.py [hunk 5]
```diff
@@ -120,6 +140,29 @@ def test_it_ignores_jobs_beyond_limit(
...
+            annotation_ids, tag="test_tag", schedule_in=MINUS_FIVE_MINUTES, force=True
+        )
+
+        queue.sync(LIMIT)
+
+        assert str({FORCED: 10}) in caplog.text
+        assert db_session.query(Job).all() == []
+        batch_indexer.index.assert_called_once_with(
+            Any.list.containing(annotation_ids).only()
+        )
```
Comment (jon-betts) [inline tests/h/services/search_index/_queue_test.py:164]: Do we want to add any testing here about the fact we don't pull these ones back from ES and Postgres for comparison?

File: h/services/search_index/_queue.py [hunk 6]
```diff
@@ -26,11 +27,11 @@ def __init__(self, db, es, batch_indexer):
         self._es = es
         self._batch_indexer = batch_indexer
 
-    def add(self, annotation_id, tag, schedule_in=None):
+    def add(self, annotation_id, tag, schedule_in=None, force=False):
         """Queue an annotation to be synced to Elasticsearch."""
```
Comment (jon-betts) [inline h/services/search_index/_queue.py:31]: Getting to the point where it's worth documenting the params here?

**Review by jon-betts — APPROVED** [review]
This fits together much more easily than I'd imagined, it just kind of does the right thing and saves us work.

I think what we have here is fine, but there are a few things I'd suggest if we want to make it better:

 * I don't think the tests demonstrate that we _don't_ poll ES and Postgres in the case of force
 * This is a nice property, and it would be good to lock it in
 * Might be nice to add param strings for the `add()` methods, which are onto their 4th arg now

File: tests/h/services/search_index/_queue_test.py (hunk 5 above)
Comment (seanh) [reply tests/h/services/search_index/_queue_test.py:164]: I think I'm gonna pass on this one, as I can't think of a good way to test it. Given that the tests use the real Postgres and Elastic, how can you test that it _doesn't_ query them? It's not that it doesn't query the DB at all either, since it does query it to get the jobs, it's that it doesn't do the annotations query. I could replace Postgres and Elastic with mocks for one test but those APIs are super ugly to mock. I think you might have to do it by patching the `_get_annotations_from_db()` and `_get_annotations_from_es()` methods but that'd couple the tests to the implementation. I can't think of a good way
